[TECH] Améliorer le temps d'exécution du processus d'anonymisation d'un utilisateur (PIX-5951)

igorissen opened the pull request:
## :jack_o_lantern: Problème

Actuellement l'anonymisation d'un utilisateur peut prendre beaucoup de temps. Ce problème de performance viendrait du fait que l'on scanne **TOUTES** les entrées de Redis pour trouver tous les refresh tokens encore valides de l'utilisateur. Ce problème de performance a été remonté par le support.

## :bat: Proposition

Après analyse, voici les solutions qui ont été évoqué :

1. Avoir une clef `user-refresh-tokens:{userid}` avec une liste des refresh tokens associé. Cette clef peut  avoir une expiration qu'on mettrait a jour a chaque rafraichissement de refresh_token. ✅
    - Éviter de faire un scan sur toutes les entrées de Redis pour trouver la liste des refresh tokens d'un utilisateur.
    - Enregistrer la liste des refresh tokens de l'utilisateur dans une nouvelle entrée pouvant accueillir une liste d'éléments, `user-refresh-tokens:{userid}`.
2. Avoir une clef `user-to-disconnect:{userId}` a true + expiration correspondant au délai d'expiration du refresh token. Si présent au moment du rafraichissement, on ne renouvelle pas le refresh_token. ❌
    - Les refresh tokens seront toujours présent jusqu'à expiration
3. Tout déplacer dans postgres avec `refresh_token, user_id, date`. ❌

## :spider_web: Remarques

Lors de la mise en production, il faudra après un délai de 7 jours environ supprimer l'ancien code qui est toujours présent pour des soucis de compatibilité. C'est à dire supprimer ces lignes ci : 
```javascript
  const userRefreshTokens = (await userRefreshTokensTemporaryStorage.lrange(userId)) || [];
  const refreshTokenFound = userRefreshTokens.find((userRefreshToken) => userRefreshToken === refreshToken);

  if (!refreshTokenFound) {
    await userRefreshTokensTemporaryStorage.lpush({ key: userId, value: refreshToken });
  }

```
```javascript
  await refreshTokenTemporaryStorage.deleteByPrefix(_prefixForUser(userId));
```

Une nouvelle notion a été rajouté dans cette PR : les listes de Redis. 

La notion de liste ajoutée dans cette PR, permet uniquement de faire le lien entre un userId et un refreshToken. 
_Pour rappel, dans redis il n'est pas possible de supprimer une entrée sans avoir la clé du-dit élément. Or sur Pix Admin nous n'avons que le userId et pas la valeur du refresh token, qui fait partie de la clé. Et, lors de la suppression d'un refresh token, on devait donc scanner / parcourir TOUTES les clefs redis, ce qui prenait beaucoup de temps._ 

Pour chaque user on a créé une liste dont la clé est userId et la valeur le refreshToken. Au moment de l'anonymisation, pour supprimer le/les refresh token de l'utilisateur on va donc : 
- récupérer la liste du userId en question, ce qui permet d'avoir tous les refresh token de l'utilisateur
- supprimer la-dite liste
- supprimer chacun des refresh token de l'utilisateur

Les éléments dans une liste redis ne sont pas périssables, c'est à dire qu'ils n'expirent jamais à moins qu'une action ne les supprime (soit au moment d'une déconnexion manuelle, ou bien une anonymisation). 
**Ce qui pose des questions sur la maintenabilité d'une liste qui ne fait que grossir (surtout dans le cas où les utilisateurs ne se déconnectent pas manuellement).** 

## :ghost: Pour tester

#### Multiples connexions + anonymisation

- Se connecter avec le même compte sur Pix app avec différents navigateurs (Chrome, Safari, Firefox, Opera, Brave, Edge, ...)
- Se connecter sur Redis
- Constatez qu'il existe différentes entrées pour chacune des connexions mais aussi une nouvelle entrée `user-refresh-tokens:<votre-user-id>` étant une liste de tous les refresh tokens de l'utilisateur
- Se connecter sur Pix Admin
- Ouvrir la page de détails de l'utilisateur
- Anonymiser cet utilisateur
- Constatez sur Redis que toutes les entrées lié à l'utilisateur ne sont plus présentes, ainsi que la nouvelle entrée `user-refresh-tokens:<votre-user-id>`

#### Une seule connexion + déconnexion

- Se connecter avec le même compte sur Pix app
- Se connecter sur Redis
- Constatez qu'il existe différentes entrées pour la connexion de l'utilisateur comprenant une nouvelle entrée `user-refresh-tokens:<votre-user-id>` étant une liste de tous les refresh tokens de l'utilisateur
- Se déconnecter
- Constatez sur Redis que toutes les entrées lié à l'utilisateur ne sont plus présentes, ainsi que la nouvelle entrée `user-refresh-tokens:<votre-user-id>`

Anne-Gaelle-S committed:
api:sr: fix typo in redis tests

igorissen committed:
feat(api): add lpush method to temporary storage

francois2metz committed:
Rename variable

francois2metz committed:
Revoke the refresh token remove it from the user id list of refresh tokens

francois2metz committed:
Revoke all refresh tokens for a user

igorissen committed:
feat(api): add more methods for redis client

lpush, lrem and lrange to handle a list of values for a key

igorissen committed:
feat(api): add more methods temporary storage

lpush, lrem and lrange to handle a list of values for a key

igorissen committed:
feat(api): override new methods of temporary storage

lpush, lrem and lrange to handle a list of values for a key

pix-service commented:
I'm deploying this PR to these urls:

- App (.fr): https://app-pr5091.review.pix.fr
- App (.org): https://app-pr5091.review.pix.org
- Orga: https://orga-pr5091.review.pix.fr
- Certif: https://certif-pr5091.review.pix.fr
- Admin: https://admin-pr5091.review.pix.fr
- API: https://api-pr5091.review.pix.fr/api/

 Please check it out!

francois2metz commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,51 @@
 }
 
 async function createAccessTokenFromRefreshToken({ refreshToken }) {
-  const { userId, source } = (await temporaryStorage.get(refreshToken)) || {};
```
Comme on ne regénère pas (encore) de refresh token au moment de donner un nouveau token, ne devrait ton pas s'assurer que le refresh token en question est bien dans la liste des user refresh token? J'ai l'impression que sinon on va perdre tout les refresh token précédent.

igorissen replied:
On pourrait le vérifier. Par contre, est-ce que ce ne sera pas un double emploi ? Dans le sens où si il existe une entrée redis pour ce refreshToken, il existera également une entrée dans la liste des refreshTokens.

francois2metz replied:
Pas forcément pour les refresh token existants en production.

Anne-Gaelle-S commented:
**Suggestions :**
- Sur la description de la PR la partie `Après analyse, voici les solutions qui ont été évoqué :` et ce qui suit devrait apparaître dans la partie # Solution
- Expliquer pourquoi on a choisis la solution 1 VS les autres solutions (pour postgres je pense avoir mon idée mais par rapport à la solution 2 je ne vois pas la différence)
- `il faudra après un délai de 7 jours environ supprimer l'ancien code ` -> est-ce que le ticket jira est déjà créé pour pas qu'on oublie ? On peut déjà préparer la PR et la mettre en "BLOCKED" tant que la date dites n'est toujours pas dépassée

igorissen committed:
feat(api): add user refresh token to a list

when creating a new access token

Anne-Gaelle-S commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
 const tokenService = require('./token-service');
 const { UnauthorizedError } = require('../../application/http-errors');
-const temporaryStorage = require('../../infrastructure/temporary-storage').withPrefix('refresh-tokens:');
+const refreshTokenTemporaryStorage = require('../../infrastructure/temporary-storage').withPrefix('refresh-tokens:');
+const userRefreshTokensTemporaryStorage = require('../../infrastructure/temporary-storage').withPrefix(
+  'user-refresh-tokens:'
+);
```
**question**
Si je comprends bien on a un renommage de `temporaryStorage` en `refreshTokenTemporaryStorage`. Et on a une nouvelle "liste" dans redis, le `userRefreshTokensTemporaryStorage` préfixée par `user-refresh-tokens:` qui lie les userId aux refresh-token.
Est-ce qu'après les 7 jours de rétrocompatibilité, on va supprimer la liste `refreshTokenTemporaryStorage` ?

igorissen replied:
C'est ça. Le processus actuel est gardé et on a ajouté une nouvelle entrée par userId qui aura une liste de refreshTokens du userId. Ceci nous permettra de récupérer la liste sans devoir faire un scan.

Oui, il faudra retirer le code lié au scan après 7 jours (environ)

Anne-Gaelle-S commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
 }
 
-async function createRefreshTokenFromUserId({ userId, source }) {
+async function createRefreshTokenFromUserId({ userId, source, uuidGenerator = uuidv4 }) {
```
**question**
On met `uuidGenerator` en paramètre pour tester plus simplement ? Ou il y a une autre raison ?

igorissen replied:
Le test actuel ne fait que vérifier que le format est le bon mais on ne vérifie pas que la valeur est la même. Pour le faire, injecter le générateur UUID permet de vérifier que la valeur est la bonne, ce qui est plus fiable que le format.

lego-technix commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
 }
 
-async function createRefreshTokenFromUserId({ userId, source }) {
+async function createRefreshTokenFromUserId({ userId, source, uuidGenerator = uuidv4 }) {
```
Je trouve ça dommage de faire cette injection de dépendance sur `uuidGenerator` car ce n'est pas une injection de dépendance « métier », c'est uniquement pour les besoins d'un test. Cela nuit à la compréhension métier de la fonction.

Ne pourrait-on pas rendre le test moins exigeant au niveau du stubbing et tester uniquement le format de la valeur de retour de `createRefreshTokenFromUserId` ?

igorissen replied:
Le test actuel ne fait que vérifier que le format est le bon mais on ne vérifie pas que la valeur est la même. Pour le faire, injecter le générateur UUID permet de vérifier que la valeur est la bonne, ce qui est plus fiable que le format.

lego-technix replied:
C'est clairement plus fiable que tester juste le format, mais c'est dommage 😢

lego-technix replied:
Il y a une solution pour stuber `uuidv4` dans la version récente du paquet `uuid` que je pense je privilégierais mais apparemment un peu hacky (je n'ai pas eu le temps de tester) donc quand même pas très satisfaisante non plus.

Anne-Gaelle-S commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
+  await bluebird.mapSeries(refreshTokens, (refreshToken) => {
+    return refreshTokenTemporaryStorage.delete(refreshToken);
+  });
+  await refreshTokenTemporaryStorage.deleteByPrefix(_prefixForUser(userId));
```
**question**
C'était pas justement cette ligne qui était très lente `await refreshTokenTemporaryStorage.deleteByPrefix(_prefixForUser(userId));` ? 
Si oui, ça veut dire que ces modifications ne répondent pas tout de suite à la problématique de lenteur sur l'anonymisation ?

igorissen replied:
C'est bien ça. On doit garder ce code pendant un moment le temps de la migration pour pouvoir le retirer par la suite.

Anne-Gaelle-S replied:
Est-ce que ça ne vaudrait pas le coup de supprimer quand même cette ligne pour rendre la vie plus facile au support pendant 7 jours ? Quitte à ne pas supprimer pendant 7jours les anciens refresh tokens (qui vont de toute manière finir par périmer). Le risque étant qu'un utilisateur revienne vers nous en disant qu'il a demandé une anoymisation et qu'il est toujours connecté mais le support peut toujours lui dire qu'on est dans un état exceptionnel et qu'il suffira à cette personne d'attendre les 7jours ?

igorissen replied:
Discuté avec l'équipe Accès. Nous allons garder les 2 processus de suppression pendant 7 jours avant de retirer l'ancienne version (le scan de toute la base de données) au profit de la liste par utilisateur.

Anne-Gaelle-S commented on api/lib/infrastructure/temporary-storage/InMemoryTemporaryStorage.js:
```diff
@@ -37,6 +37,25 @@ class InMemoryTemporaryStorage extends TemporaryStorage {
     noop;
   }
+
+  lpush(key, value) {
```
**question**
Je ne comprends pas ces nommages, `lpush`, `lrem`, `lrange`. Ca sort d'où ?

igorissen replied:
En faite c'est le nom des méthodes de redis que l'on a utilisé. J'ai du mal a trouver un nom correcte pour ces méthodes.

lego-technix replied:
Quand on connait ou qu'on veut se documenter sur Redis c'est très clair et parfaitement documenté : https://redis.io/commands/. On pourrait éventuellement nommer ces fonctions différemment mais cela représente un travail qui demanderait du temps pour trouver quelque chose d'uniforme alors que là nous avons un nommage qui se base sur de l'existant connu et documenté.

Anne-Gaelle-S commented on api/lib/infrastructure/temporary-storage/InMemoryTemporaryStorage.js:
```diff
@@ -37,6 +37,25 @@ class InMemoryTemporaryStorage extends TemporaryStorage {
+  }
+
+  lrange(key) {
+    return this._client.get(key) || [];
```
**question**
C'est juste un `get` qui renvoie un tableau vide ... j'ai du mal à comprendre l'intérêt de cette fonction. Ca doit être en lien avec ma question précédente mais du coup pourquoi avoir ajouté cette fonction plutôt qu'utiliser un `get` classique ?

igorissen replied:
Comme c'est le InMemoryTemporaryStorage, on peut se permettre de faire ça. De plus on est obligé d'override cette méthode.

Anne-Gaelle-S commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
-    key: token,
+  const refreshToken = `${_prefixForUser(userId)}${uuidGenerator()}`;
+
+  await refreshTokenTemporaryStorage.save({
```
**question**
On est pas sensé supprimer cette partie là ? 
On ne veut plus générer les anciens refresh token dès cette PR non ?

igorissen replied:
Non. On doit garder cette ligne. Il faut une entrée par userId et par refreshToken pour pouvoir bénéficier de l'expiration/suppression automatique de l'entrée. La liste est là juste pour supprimer potentiellement des entrées qui sont encore disponibles.

Anne-Gaelle-S commented:
Sur Pix App je tombe sur une boucle infinie même si j'ai `REFRESH_TOKEN_LIFESPAN=30s` je ne suis jamais déconnectée : il me génère des access-tokens à l'infinis
<img width="1446" alt="image" src="https://user-images.githubusercontent.com/38167520/196917600-e5637c3c-80b8-4cd0-92f4-fca2bc8caa90.png">

Voilà les logs redis en lien : 
<img width="956" alt="image" src="https://user-images.githubusercontent.com/38167520/196918005-009306f5-6408-402d-a2fb-a58b8a85db41.png">

francois2metz commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
     value: { type: 'refresh_token', userId, source },
     expirationDelaySeconds,
   });
+  await userRefreshTokensTemporaryStorage.lpush({ key: userId, value: refreshToken });
```
Nous avons probablement oublié de mettre l'expiration de cette clef ici. A mettre probablement a la même valeur que le refresh token (+1 heure ou un truc du genre). De cette façon on s'assure de nettoyer les clefs d'un utilisateur qui ne se connecte que très rarement.

igorissen replied:
Done

francois2metz commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,59 @@
+  const refreshTokenFound = userRefreshTokens.find((userRefreshToken) => userRefreshToken === refreshToken);
+
+  if (!refreshTokenFound) {
+    await userRefreshTokensTemporaryStorage.lpush({ key: userId, value: refreshToken });
```
L'expiration sera également a mettre à jour ici.

igorissen replied:
Done

igorissen committed:
feat(api): add an expiration delay to a user refresh tokens list

igorissen committed:
feat(api): update user refresh tokens list expiration delay

when creating a new access token from refresh token and the refresh token is not in the list (compatibility with previous code version)

francois2metz commented on api/lib/domain/services/refresh-token-service.js:
```diff
@@ -1,35 +1,74 @@
+
+  const refreshTokenRemainingExpirationSeconds = await refreshTokenTemporaryStorage.ttl(refreshToken);
+  const userRefreshTokenRemainingExpirationSeconds = await userRefreshTokensTemporaryStorage.ttl(refreshToken);
+  const userRefreshTokens = (await userRefreshTokensTemporaryStorage.lrange(userId)) || [];
```
A ma connaissance, `lrange` revoit toujours un tableau coté redis.

francois2metz reviewed (approved):
:leafy_green: :maple_leaf: :four_leaf_clover: :fallen_leaf:  LGTM :fallen_leaf:  :four_leaf_clover: :maple_leaf: :leafy_green:
